### PR TITLE
WIP Merging Changes

### PR DESCRIPTION
**Pull request by slinksoft**
In progress (still working on unit and integration tests)

### CODE REVIEW

File: exactdelaypathfinder/core.py [hunk 1]
```diff
@@ -39,14 +55,25 @@ def search(self, graph, total_delay, start, end, max_results=10):
...
+           for the exact and closest paths based upon user-input propagation delay.
+           This recursive function is called in the "search" function.
+
+           Parameters:
+              delay: the current delay that has been traversed so far (the function is recursive so think of it in that sense)
+              curr: The current node for the path we are on
+              target: The target node for the path we are making
+              path: the current path we have traversed
+              visits: Used to restrict the amount of visits on a node that has been previously traversed
+        """
```
Comment (MarvinTorres) [inline exactdelaypathfinder/core.py]: Private methods should not have docstrings.
```suggestion
```

File: tests/test_core.py [hunk 2]
```diff
@@ -1,10 +1,95 @@
...
+        """
+        new_mock_EDPF = mock_EDPF.return_value
+        new_mock_EDPF._search.return_value = 86
+        G = Mock()
+        delay = 64;
+        start = 'a'
+        end = 'b'
+        new_mock_EDPF._search(G, 64, 'a', 'b', [])
+        new_mock_EDPF._search.assert_called_with(G, delay, start, end, visits=[])
+
```
Comment (MarvinTorres) [inline tests/test_core.py]: This is how to mock a method without mocking the entire class:
```suggestion
    @patch.object(ExactDelayPathfinder, '_search')
    def test_search(self, mock_search):
        """ Tests the _search function, the DFS algorithm (recursive method) 
        
            Parameters:
               mock_EDPF: Mock object reference of ExactDelayPathfinder
        """
        new_EDPF = ExactDelayPathfinder()
        mock_search.return_value = []

        G = Mock()
        G.nodes = []
        delay = 64;
        start = 'a'
        end = 'b'
        new_EDPF.search(G, 64, 'a', 'b')
        new_EDPF._search.assert_called()
```

File: tests/test_core.py [hunk 3]
```diff
@@ -1,10 +1,95 @@
...
+    @patch('exactdelaypathfinder.core.ExactDelayPathfinder', autospec=True)
+    def test_search_delay_Error(self, mock_EDPF):
+        """ Tests the search function attribute errors for the delay input parameter 
+        
+            Parameters:
+               mock_EDPF: Mock object reference of ExactDelayPathfinder
+        """
+        new_mock_EDPF = ExactDelayPathfinder()
+        G = Mock()
+        self.assertRaises(AttributeError, new_mock_EDPF.search, G, -10,'a', 'b')
```
Comment (MarvinTorres) [inline tests/test_core.py]: No need for mocking here.

File: tests/test_core.py [hunk 4]
```diff
@@ -1,10 +1,95 @@
...
+    @patch('exactdelaypathfinder.core.ExactDelayPathfinder', autospec=True)
+    def test_search_max_result_Error(self, mock_EDPF):
+        """ Tests the search function attribute errors for the max result parameter 
+        
+            Parameters:
+               mock_EDPF: Mock object reference of ExactDelayPathfinder
+        """
+        new_mock_EDPF = ExactDelayPathfinder()
+        G = Mock()
+        self.assertRaises(AttributeError, new_mock_EDPF.search, G, -10,'a', 'b', -5)
```
Comment (MarvinTorres) [inline tests/test_core.py]: Also here.

**Review by MarvinTorres — CHANGES_REQUESTED** [review]
Hi Daniel, I have given suggestions on how to improve the test code. Let me know what you think.

**Comment by slinksoft** [comment]
Commited all your suggestions. Everything should be good now. Now all we need to do is create the integration test suite. Should be able to get that done by the end of the week!

**Comment by MarvinTorres** [comment]
This is perfect! I will see if I can squash your changes then merge the squash.

**Pull request merged**